warehouse/oidc/models.py

```diff
@@ -0,0 +1,48 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from sqlalchemy import Column, ForeignKey, String
+from sqlalchemy.dialects.postgresql import UUID
+
+from warehouse import db
+
+
+class OIDCProvider(db.Model):
+    __tablename__ = "oidc_providers"
+
+    discriminator = Column(String)
+
+    __mapper_args__ = {"polymorphic_on": discriminator}
+
+    def verify_claims(self, token):
+        return NotImplemented
+
+
+class GitHubProvider(OIDCProvider):
+    __tablename__ = "github_oidc_providers"
+    __mapper_args__ = {"polymorphic_identity": "GitHubProvider"}
+
+    id = Column(UUID(as_uuid=True), ForeignKey(OIDCProvider.id), primary_key=True)
+    repository_name = Column(String)
+    owner = Column(String)
+    owner_id = Column(String)
+    workflow_name = Column(String)
+
+    def repository(self):
+        return f"{self.owner}/{self.repository_name}"
+
+    def job_workflow_ref(self):
+        return f"{self.repository}/.github/workflows/{self.workflow_name}.yml"
```

Comment:
cc @di: How clever do you want us to get with the "claims" here?

I could add a `@claim` decorator that transparently updates the instance's `self.claims` list as you suggested in your example, but I wasn't sure whether that amount of magic was worth it for what'll end up being just a handful of methods.

Reply:
I mostly just to make it easy to verify all claims and make it hard to forget to verify some claim.

Regarding the decorator, it'd probably be just as likely that we forget to add it as it would be that we forget to verify the claim.

Maybe something like explicit like `__verified_claims__ == {'repository', 'job_workflow_ref', ...}` and `__unverified_claims__ == {'<something else>'}` and we ensure that every claim in the token is either explicitly verified or unverified? And in the case where there's a claim that's unaccounted for (either it's new or we aren't verifying it) an appropriate warning is generated? (probably not an outright failure b/c we probably don't want to fail on every new claim?)

Reply:
Sounds good to me!

Reply:
`__unverified_claims__` will have to include the set of "standard" claims that are verified during signature verification, but we can document that explicitly.

Reply:
Okay, this is what I've come up with:

```python
    __verified_claims__ = {
        "repository",
        "job_workflow_ref",
        "actor",
        "workflow",
    }

    __preverified_claims__ = {
        "iss",
        "iat",
        "nbf",
        "exp",
        "aud",
    }
    
    __unverified_claims__ = {
        "jti",
        "sub",
        "ref",
        "sha",
        "run_id",
        "run_number",
        "run_attempt",
        "head_ref",
        "base_ref",
        "event_name",
        "ref_type",
    }
```

We need to think some more about how we actually do the verification step for each claim: most claims are just a string equality check, but some are prefix checks instead (like `job_workflow_ref`). Do we want to something dynamic like fetching a `verify_{claim}` function for each claim and calling that? That's pretty magical, so I'm not sure if that's the best approach.

Reply:
Maybe each claim could just be `@property` on the class? So verification would be something like:

```python
def verify_claims(self, signed_token):
    for claim_key in self.__verified_claims__:
        if not getattr(self, claim_key) == signed_token[claim_key]:
            return False
    return True
```

Reply:
That's pretty close to what we want, although we need more complex behavior than just equality checking. Perhaps something like this?

```python
def verify_claims(self, signed_token):
    for claim_key in self.__verified_claims__:
        if not getattr(self, claim_key)(signed_token[claim_key]):
            return False
    return True
```

i.e., each would be a function instead of a property, allowing us to do e.g. substring matching.